### PR TITLE
tls: Remove tls slot callback replacement-value

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Removes the return-value semantics for runOnAllThreads callbacks; now the function type has void return, so we can remove the ASSERTs and simplify the code.

Also changes a few more instances of SlotPtr to TypedSlot. Still about 30 remain.

Moves Slot::runOnAllThreads to the protected section to prevent new dynamically uses from emerging.
Additional Description:
Risk Level: low
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: include/envoy/thread_local/thread_local.h [hunk 1]
```diff
@@ -76,19 +76,14 @@ class Slot {
...
   /**
-   * UpdateCb takes the current stored data, and must return the same data. The
-   * API was designed to allow replacement of the object via this API, but this
-   * is not currently used, and thus we are removing the functionality. In a future
-   * PR, the API will be removed.
-   *
-   * TLS will run the callback and assert the returned returned value matches
-   * the current value.
+   * UpdateCb takes is passed a shared point to the current stored data. Use of
+   * this API is deprecated; please use TypedSlot::runOnAllThreads instead.
```
Comment (mattklein123) [inline include/envoy/thread_local/thread_local.h:80]: Can we just make this private and use a friend declaration or something like that?

**Review by mattklein123 — COMMENTED** [review]
Awesome! One small question.

/wait-any

File: include/envoy/thread_local/thread_local.h (hunk 1 above)
Comment (jmarantz) [reply include/envoy/thread_local/thread_local.h:85]: protected, yeah. I had been thinking about that but got lazy. Done now. Doing so made it necessary to move the testing to the TypedSlot wrapper, but i think that makes sense.

It also exposed some users of Slot::runOnAllThreads that I had somehow missed earlier, so those needed to be cleaned up.

File: .bazelversion [hunk 2]
```diff
@@ -1 +1 @@
-3.6.0
+3.7.0
```
Comment (mattklein123) [inline .bazelversion:1]: ?

**Review by mattklein123 — APPROVED** [review]
Nice!

File: .bazelversion (hunk 2 above)
Comment (mattklein123) [reply .bazelversion:1]: (Sorry didn't mean to approve, I think this needs to be reverted.)

/wait

Comment (jmarantz) [reply .bazelversion]: done; didn't mean to commit that. Having some trouble with fuzz-coverage tests anyway; not sure what the root cause is. The log didn't help me but I'll ask @asraa for help tomorrow.

Comment (jmarantz) [reply .bazelversion]: fuzz_coverage passed this time; must be an infrastructural flake.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**